src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -52,67 +76,176 @@ public void AddDependencyLocation(string fullPath)
-            if (loadedAssembly == null)
+            string realPath = PreparePathToLoad(originalAnalyzerPath);
+            AssemblyName? assemblyName;
+            try
             {
-                loadedAssembly = LoadFromPathUncheckedImpl(fullPath);
+                assemblyName = AssemblyName.GetAssemblyName(realPath);
+            }
+            catch
+            {
+                // The above can fail with the assembly doesn't exist because it's corrupted, 
+                // doesn't exist on disk or is a native DLL. Those failures are handled when 
+                // the actual load is attempted. Just record the failure now.
+                assemblyName = null;
             }
 
-            // Add the loaded assembly to the path cache.
             lock (_guard)
             {
-                _loadedAssembliesByPath[fullPath] = loadedAssembly;
+                _analyzerAssemblyInfoMap[originalAnalyzerPath] = (assemblyName, realPath);
             }
 
-            return loadedAssembly;
+            return (assemblyName, realPath);
         }
 
-        protected ImmutableHashSet<string>? GetPaths(string simpleName)
+        /// <summary>
+        /// Return the best path for loading an assembly with the specified <see cref="AssemblyName"/>. This
+        /// return is a real path to load, not an original path.
+        /// </summary>
+        protected string? GetBestPath(AssemblyName assemblyName)
         {
+            if (assemblyName.Name is null)
+            {
+                return null;
+            }
+
+            ImmutableHashSet<string>? paths;
             lock (_guard)
             {
-                _knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out var paths);
-                return paths;
+                if (!_knownAssemblyPathsBySimpleName.TryGetValue(assemblyName.Name, out paths))
+                {
+                    return null;
+                }
             }
+
+            // Sort the candidate paths by ordinal, to ensure determinism with the same inputs if you were to have
+            // multiple assemblies providing the same version.
+            string? bestPath = null;
+            AssemblyName? bestName = null;
+            foreach (var candidateOriginalPath in paths.OrderBy(StringComparer.Ordinal))
+            {
+                (AssemblyName? candidateName, string candidateRealPath) = GetAssemblyInfoForPath(candidateOriginalPath);
+                if (candidateName is null)
+                {
+                    continue;
+                }
+
+                bool isMatch;
+#if NETCOREAPP
+                isMatch = candidateName.Name == assemblyName.Name;
+#else
+                isMatch =
+                    candidateName.Name == assemblyName.Name &&
+                    candidateName.Version >= assemblyName.Version &&
+                    candidateName.GetPublicKeyToken().AsSpan().SequenceEqual(assemblyName.GetPublicKeyToken().AsSpan());
```

Comment:
>candidateName.GetPublicKeyToken().AsSpan().SequenceEqual(assemblyName.GetPublicKeyToken().AsSpan() [](http://example.com/codeflow?start=20&length=98)

Is the public key check not needed with `NETCOREAPP`?

Reply:
On .NET Core the only thing that matters is simple name. The runtime never considers public key tokens or versions. Thus our dependency loading has to do the same.